TIP-1064: StablecoinDEX order storage credits (#4082)
tempoxyz/tempo: tips/tip-1064.md

Without user-level accounting, a user who creates new order state could consume storage gas tokens minted by another user's order deletion. However, attributing shared accounting storage to the user whose operation happens to clear it can create windfall credits. For example, the user who removes the last order at a tick should not receive an extra user credit merely because the DEX can clear a shared tick-level record or bitmap word.

This TIP creates the user-level accounting suggested by TIP-1058 for the StablecoinDEX only.

---

# Specification

## Terminology

- **DEX storage credit**: A protocol-maintained `uint64` counter associated with one user address in the StablecoinDEX.
- **Reusable order storage**: StablecoinDEX-owned storage that is clearly attributable to one order for storage creation and deletion accounting.
- **Storage owner**: The user address charged or credited for a reusable order storage slot.
- **TIP-1060 token**: A precompile-level storage gas token minted or consumed by the StablecoinDEX under TIP-1060.

## State

The StablecoinDEX MUST maintain:

```text
dex_storage_credits: mapping(address user => uint64 credits)
```

> The `dex_storage_credits` mapping is StablecoinDEX precompile state. Creating or updating a user's `dex_storage_credits` counter, including the first creation of that counter's backing storage slot, is subject to normal TIP-1000 state gas and TIP-1060 rules.

Reviewer: ```suggestion
The `dex_storage_credits` mapping is StablecoinDEX precompile state. Creating or updating a user's `dex_storage_credits` counter, including the first creation of that counter's backing storage slot, is subject to normal TIP-1000 state gas and TIP-1060 rules. The StablecoinDEX MUST consume a TIP-1060 storage token to change the `dex_storage_credits[U]` array from `0` to `1`. Inversely it will obtain a TIP-1060 storage credit when `dex_storage_credits[U]` returns to its zero value. One of the user `U`'s storage credit is effectively stored in the slot occupied by the `dex_storage_credits[U]` value.
```

Only storage owned by the StablecoinDEX is covered. TIP-20 token balances, allowances, policy state, EOA state, and ordinary smart-contract storage are outside this TIP's scope.

## Storage Owner Attribution

For each reusable order storage slot, the storage owner is:

| Storage | Storage owner |
|---------|---------------|
| Order record | `order.maker` |
| Client order ID mapping | the maker of the order identified by the mapping entry |

When a single operation touches storage owned by multiple orders, each reusable order storage creation or deletion MUST be attributed independently using this table.

Storage that is not listed in this table MUST NOT mint, consume, credit, or debit DEX storage credits for any user. If a storage slot cannot be attributed to one order, it is non-creditable accounting storage even if the operation that creates or deletes it was initiated by a known user.

> ## Crediting Deleted Storage

Reviewer 2: > 2. Active client order ID mappings keyed by `(user, clientOrderId)`, if TIP-1003 is active and the mapping entry identifies one active order.

For clarity, would it be technically possible for a user to abuse this by placing `n` orders and then immediately deleting them when fees are exceptionally low. Then if fees increaese, when placing new orders, they are effectively subsidized at a much higher price than what they actually paid.

## Query Interface

The StablecoinDEX interface MUST expose:

```solidity
/// @notice Returns the number of reusable order storage credits owned by a user.
/// @param user The user whose DEX storage credit balance is queried.
/// @return credits The number of storage credits available to that user.
function storageCredits(address user) external view returns (uint64 credits);
```

This function is read-only and does not expose any ability to transfer, mint, burn, or set storage credits.

## Compatibility

This TIP changes gas costs for StablecoinDEX calls that create reusable order storage. A user who has previously freed order-attributable DEX storage may pay less for later order-attributable DEX storage creation. A user without credits still pays the TIP-1000 storage creation cost for net-new order storage.

Existing DEX functional behavior, matching behavior, order priority, balances, and events are unchanged except for the new `storageCredits(address)` view.

---

# Invariants

> 

Reviewer 3: Is the invariant 
  TIP1060.balanceOf(STABLECOIN_DEX_ADDRESS) >= sum(StablecoinDEX.storageCredits(user))
or 
  TIP1060.balanceOf(STABLECOIN_DEX_ADDRESS) == sum(StablecoinDEX.storageCredits(user))
?

I think the former because it switches to preserve mode sometimes

Reviewer 4: it is the former, cause some credits are not attributed to any users

Without user-level accounting, a user who creates new order state could consume storage gas tokens minted by another user's order deletion. However, attributing shared accounting storage to the user whose operation happens to clear it can create windfall credits. For example, the user who removes the last order at a tick should not receive an extra user credit merely because the DEX can clear a shared tick-level record or bitmap word.

This TIP creates the user-level accounting suggested by TIP-1058 for the StablecoinDEX only.

---

# Specification

## Terminology

- **DEX storage credit**: A protocol-maintained `uint64` counter associated with one user address in the StablecoinDEX.
- **Reusable order storage**: StablecoinDEX-owned storage that is clearly attributable to one order for storage creation and deletion accounting.
- **Storage owner**: The user address charged or credited for a reusable order storage slot.
- **TIP-1060 token**: A precompile-level storage gas token minted or consumed by the StablecoinDEX under TIP-1060.

## State

The StablecoinDEX MUST maintain:

```text
dex_storage_credits: mapping(address user => uint64 credits)
```

> The `dex_storage_credits` mapping is StablecoinDEX precompile state. Creating or updating a user's `dex_storage_credits` counter, including the first creation of that counter's backing storage slot, MUST NOT itself mint or consume TIP-1060 tokens and MUST NOT incur the 250,000 gas TIP-10-mandated storage creation cost.

Reviewer: ```suggestion
The `dex_storage_credits` mapping is StablecoinDEX precompile state. Creating or updating a user's `dex_storage_credits` counter, including the first creation of that counter's backing storage slot, is subject to normal TIP-1000 state gas and TIP-1060 rules.
```

Only storage owned by the StablecoinDEX is covered. TIP-20 token balances, allowances, policy state, EOA state, and ordinary smart-contract storage are outside this TIP's scope.

## Storage Owner Attribution

For each reusable order storage slot, the storage owner is:

| Storage | Storage owner |
|---------|---------------|
| Order record | `order.maker` |
| Client order ID mapping | the maker of the order identified by the mapping entry |

When a single operation touches storage owned by multiple orders, each reusable order storage creation or deletion MUST be attributed independently using this table.

Storage that is not listed in this table MUST NOT mint, consume, credit, or debit DEX storage credits for any user. If a storage slot cannot be attributed to one order, it is non-creditable accounting storage even if the operation that creates or deletes it was initiated by a known user.

## Crediting Deleted Storage

When a StablecoinDEX operation changes reusable order storage owned by user `U` from nonzero to zero:

1. Apply the normal DEX state transition.

> 2. Mint the StablecoinDEX TIP-1060 token according to TIP-1060.

Reviewer: ```suggestion
2. The TIP-1060 precompile automatically credits the StablecoinDEX with a storage token.
```

## StablecoinDEX Operation Rules

Order placement and flip-order placement MUST attribute new order records and active client order ID mappings to the order maker.

Order cancellation and full order fills MUST attribute deleted order records and deleted active client order ID mappings to the order maker.

Tick-level records, tick bitmap words, orderbook records, user internal balance slots, aggregate liquidity fields, and other DEX accounting storage MUST NOT create or delete DEX storage credits. In particular, clearing the last order at a tick MUST NOT give the order maker extra credits for clearing tick-level or bitmap storage.

Partial fills MUST NOT create or delete DEX storage credits unless they also cause reusable order storage to transition between zero and nonzero.

Internal balance updates MUST NOT create or delete DEX storage credits.

TIP-1056 flip-in-place behavior does not delete and recreate the order record when a flip succeeds. A successful in-place flip therefore MUST NOT credit or charge DEX storage credits for the order record itself. Tick-level or bitmap changes caused by moving the order are non-creditable accounting storage.

> 

Reviewer: ```suggestion
```

Existing DEX functional behavior, matching behavior, order priority, balances, and events are unchanged except for the new `storageCredits(address)` view.

---

# Invariants

1. A user's DEX storage credit balance increases only when reusable order storage attributed to that user changes from nonzero to zero.
2. A user's DEX storage credit balance decreases only when reusable order storage attributed to that user changes from zero to nonzero and a StablecoinDEX TIP-1060 token is consumed.
3. A DEX storage credit can offset at most one reusable order storage creation.
4. DEX storage credits are not transferable between users.
5. StablecoinDEX TIP-1060 tokens cannot be consumed for a user who has no DEX storage credits.
6. Storage credit accounting changes revert with the execution scope that caused them.
7. Creating or updating `dex_storage_credits` counters does not mint or consume TIP-1060 tokens and does not incur the 250,000 gas TIP-10-mandated storage creation cost.
8. Ordinary smart contracts, TIP-20 token state, and non-DEX precompile state do not mint or consume DEX storage credits under this TIP.

> 9. Non-creditable accounting storage, including tick-level records, tick bitmap words, orderbook records, and internal balance slots, does not mint, consume, credit, or debit DEX storage credits.

Reviewer: ```suggestion
7. Non-creditable accounting storage, including tick-level records, tick bitmap words, orderbook records, and internal balance slots, does not mint, consume, credit, or debit DEX storage credits.
```
